undofx/src/main/java/org/fxmisc/undo/impl/UndoManagerImpl.java

```diff
@@ -209,8 +229,10 @@ private void addChange(C change) {
     @SuppressWarnings("unchecked")
     private C[] merge(C c1, C c2) {
         Optional<C> merged = merge.apply(c1, c2);
-        if(merged.isPresent()) {
-            return (C[]) new Object[] { merged.get() };
+        if(merged.isPresent() && isNotIdentity.test(merged.get())) {
+            return isNotIdentity.test(merged.get())
```

Comment:
K. I've fixed it now.
I in-lined `merge` into the method because it makes it easier to read since one has to set `canMerge` on each part.
I also split up the tests so that one doesn't interfere with another.